### PR TITLE
feat(rt): introduce TaskLocalAlloc

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

The PR introduced the `TaskLocalAlloc`, which will track the stats of memory allocation by every `tokio::Task`.

For the detail, address the comments in codes.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/utils/stats_alloc/src/task_local.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                let ptr = ptr.add(usize::BITS as usize);
+                ptr
+            })
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: std::alloc::Layout) {
+        let new_layout =
+            Layout::from_size_align_unchecked(layout.size() + usize::BITS as usize, layout.align());
+        let ptr = ptr.sub(usize::BITS as usize);
+        let bytes = ptr as *const AtomicUsize;
```
Comment (zwang28) [inline src/utils/stats_alloc/src/task_local.rs:62]: Is dangling pointer possible?

**Comment by liurenjie1024** [comment]
So this allocator can be used to collect task memory usage, what if I want to collecto per executor memory usage?

File: src/utils/stats_alloc/src/task_local.rs (hunk 1 above)
Comment (BugenZhao) [reply src/utils/stats_alloc/src/task_local.rs]: We've refactored the implementation and it'll check whether the stats is valid before mutating.

**Comment by BugenZhao** [comment]
> So this allocator can be used to collect task memory usage, what if I want to collecto per executor memory usage?

The scope of `task_local` can be smaller than a tokio task, and it also supports shallowing. For example, we can provide the scope on each future of `executor.next()` (like tracing), so that allocation of this executor during the processing of this chunk can be recorded.

File: src/utils/stats_alloc/src/task_local.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+use std::alloc::{GlobalAlloc, Layout, System};
+use std::future::Future;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::time::Duration;
+
+use tokio::task_local;
+
+#[repr(transparent)]
+#[derive(Clone, Copy)]
+struct TaskLocalBytesAllocated(Option<&'static AtomicUsize>);
```
Comment (BugenZhao) [inline src/utils/stats_alloc/src/task_local.rs:29]: If we're going to release it instead of leaking it, maybe using a `NonNull` pointer here will be more suitable?

File: src/utils/stats_alloc/src/task_local.rs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        self.0.unwrap_unchecked().fetch_add(val, Ordering::Relaxed);
+    }
+
+    #[inline(always)]
+    pub fn sub(&self, val: usize) {
+        if let Some(bytes) = self.0 {
+            let old_bytes = bytes.fetch_sub(val, Ordering::Relaxed);
+            if old_bytes - val == 0 {
+                unsafe { Box::from_raw_in(bytes.as_mut_ptr(), System) };
+            }
```
Comment (BugenZhao) [inline src/utils/stats_alloc/src/task_local.rs:66]: We're actually emulating an `Arc`, and the ordering here might be problematic. Need further investigation.

https://doc.rust-lang.org/src/alloc/sync.rs.html#1664

Comment (BugenZhao) [reply src/utils/stats_alloc/src/task_local.rs:66]: If there're adding or subtracting `0` bytes, this doesn't work as well.

There's no guarantee that the size of layout is non-zero. We may need to check this explicitly. Alternative: take the 8 byte of counter into account with `wrapped_layout.size()`.

> This function is unsafe because undefined behavior can result if the caller does not ensure that layout has non-zero size.

**Review by BugenZhao — COMMENTED** [review]
Here leaves some TODOs for this work. @TennyZhuang and I will continue working on this.

**Comment by TennyZhuang** [comment]
Hi, all, the PR is ready for review.

**Review by jon-chuang — COMMENTED** [review]
Can we see some unittest showing the expected allocation agrees with measured? Also I'm suspicious of the dealloc impl, maybe it will be good to test dealloc too?

File: src/utils/stats_alloc/src/task_local.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                *ptr.cast() = TaskLocalBytesAllocated::invalid();
+                ptr.wrapping_add(offset)
+            })
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
+        let (wrapped_layout, offset) = wrap_layout(layout);
+        let ptr = ptr.wrapping_sub(offset);
+
+        let bytes: TaskLocalBytesAllocated = *ptr.cast();
```
Comment (jon-chuang) [inline src/utils/stats_alloc/src/task_local.rs:164]: What's the purpose of this? Those bytes are going to be dealloc anyway. In the first place, why do we prefix every alloc with this type? There doesn't seem to be a point, all alloced bytes can simply be counted by the task local var. 

Of course, we can encounter the scenario where another task frees the allocated memory etc. But maybe this is out of scope.

File: src/utils/stats_alloc/src/task_local.rs [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+                ptr.wrapping_add(offset)
+            })
+    }
+
+    unsafe fn realloc(&self, ptr: *mut u8, layout: Layout, new_size: usize) -> *mut u8 {
+        let (wrapped_layout, offset) = wrap_layout(layout);
+        let ptr = ptr.wrapping_sub(offset);
+
+        let bytes: TaskLocalBytesAllocated = *ptr.cast();
+        bytes.add(new_size);
```
Comment (jon-chuang) [inline src/utils/stats_alloc/src/task_local.rs:192]: Shouldn't we apply this to the task local var instead?

File: src/utils/stats_alloc/src/task_local.rs (hunk 4 above)
Comment (jon-chuang) [reply src/utils/stats_alloc/src/task_local.rs:164]: Anw, I think we can get rid of this extra allocation and rely only on task local var. Anw, we should see if unit test passes on current and if not check if this change will pass.

Comment (BugenZhao) [reply src/utils/stats_alloc/src/task_local.rs:164]: > There doesn't seem to be a point, all alloced bytes can simply be counted by the task local var.

If we allocated some bytes in actor 1, send them to actor 2 through the channel, then actor 2 deallocates them, the substraction will be applied on the actor 2's task-local and the result becomes incorrect.

File: src/utils/stats_alloc/src/task_local.rs (hunk 5 above)
Comment (BugenZhao) [reply src/utils/stats_alloc/src/task_local.rs:192]: As above, the deallocation should always be associated with the task that originally allocates it, so we don't use task local var here.

File: src/utils/stats_alloc/src/task_local.rs [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            let old_bytes = bytes.fetch_sub(val, Ordering::Release);
+            // If the counter reaches zero, delete the counter. Note that we've ensured there's no
+            // zero deltas in `wrap_layout`, so there'll be no more uses of the counter.
+            if old_bytes == val {
+                // This fence is needed to prevent reordering of use of the counter and deletion of
+                // the counter. Because it is marked `Release`, the decreasing of the counter
+                // synchronizes with this `Acquire` fence. This means that use of the counter
+                // happens before decreasing the counter, which happens before this fence, which
+                // happens before the deletion of the counter.
+                fence(Ordering::Acquire);
```
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/utils/stats_alloc/src/task_local.rs:78]: Looks good to me!

File: src/utils/stats_alloc/src/task_local.rs (hunk 5 above)
Comment (jon-chuang) [reply src/utils/stats_alloc/src/task_local.rs:192]: so, how do we tally up the total deallocated/reallocated?

It doesn't seem to be done here.

Comment (jon-chuang) [reply src/utils/stats_alloc/src/task_local.rs:192]: oh, I see, we use a static ptr.

Anw, I'd still like to see a unittest, including multithreaded test, that shows this works as expected (including across threads).

File: src/utils/stats_alloc/tests/integration.rs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            |_| {},
+        )
+        .await;
+    }
+    // There should only one u64 held by `placeholder` not dropped in the task local allocator.
+    assert_eq!(bytes.unwrap().val(), std::mem::size_of::<u64>());
+    // The placeholder was consumed by `black_box`.
+    black_box(placeholder);
+    // The counter should be dropped, but there is no way to test it.
+}
```
Comment (jon-chuang) [inline src/utils/stats_alloc/tests/integration.rs:51]: perhaps for multithreaded test, we can spawn a new runtime (or simply spawn a new thread), move some on-heap data into the runtime, and free it from that runtime. Then we can check that in the original runtime, the allocation in fact goes back to 0.

**Comment by liurenjie1024** [comment]
Should we add tests using loom?

**Comment by BowenXiao1999** [comment]
Bench it with sysbench, draft result: https://www.notion.so/risingwave-labs/Test-Sysbench-against-TaskLocalAllocator-494685119f8a47af810cbe27631efd6c

I pick some numbers to follow: p99 latency, eps, qps. (Name more if you have better choices)

Roughly found that indeed there is performance drop, but not so high, should be controlled down to 15%. (e.g., primary key point lookup, for this branch the eps is 15172, while for main branch is 16966). 

The avg execution time seems do not differ too much. BTW, their grafana seems can not tell something useful so I do not dive deep.

I plan to do some TPC-H or Nexmark test next. Reply here if there is some important metrics to observe (Cuz there are too many..). Not so sure what should focus.

**Comment by liurenjie1024** [comment]
I think for streaming's nexmark test, the most important metrics is memory usage in compute node and source throughtput.

**Comment by BowenXiao1999** [comment]
> the most important metrics is memory usage in compute node

I also have similar thoughts. But IIUC, the memory usage should not differ cuz the two are both using Jemalloc internally and managed in same mechanism (?).

**Comment by liurenjie1024** [comment]
> > the most important metrics is memory usage in compute node
> 
> I also have similar thoughts. But IIUC, the memory usage should not differ cuz the two are both using Jemalloc internally and managed in same mechanism (?).

It's hard to tell, adding extra 8 byte to each allocation not only adding just 8 bytes, it changed the layout, so I'm not sure the effect, and we need metrics for it.

File: src/utils/task_stats_alloc/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+unsafe impl<A> GlobalAlloc for TaskLocalAlloc<A>
+where
+    A: GlobalAlloc,
+{
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let (wrapped_layout, offset) = wrap_layout(layout);
+
+        BYTES_ALLOCATED
+            .try_with(|&bytes| {
+                bytes.add_unchecked(layout.size());
```
Comment (fuyufjh) [inline src/utils/task_stats_alloc/src/lib.rs:147]: I think the aligned size is more accurate.

```suggestion
                bytes.add_unchecked(layout.align());
```

Comment (BugenZhao) [reply src/utils/task_stats_alloc/src/lib.rs:147]: I guess the `align` is not what we want? 🤔

> The minimum byte alignment for a memory block of this layout.

> `size`, when rounded up to the nearest multiple of `align`,

**Comment by BowenXiao1999** [comment]
Update results for nexmark q7: https://www.notion.so/risingwave-labs/Test-Sysbench-against-TaskLocalAllocator-494685119f8a47af810cbe27631efd6c.

The streaming source throughput indeed dropped a lot (main: 1579418, this pr: 955210). Because the throughput is droppped, so the memory usage do not go up too much (main: 9.21GB, this pr: 7.45GB).

May need to find more test.

File: src/utils/task_stats_alloc/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+            let old_bytes = bytes.fetch_sub(val, Ordering::Release);
+            // If the counter reaches zero, delete the counter. Note that we've ensured there's no
+            // zero deltas in `wrap_layout`, so there'll be no more uses of the counter.
+            if old_bytes == val {
+                // This fence is needed to prevent reordering of use of the counter and deletion of
+                // the counter. Because it is marked `Release`, the decreasing of the counter
+                // synchronizes with this `Acquire` fence. This means that use of the counter
+                // happens before decreasing the counter, which happens before this fence, which
+                // happens before the deletion of the counter.
+                fence(Ordering::Acquire);
```
Comment (liurenjie1024) [inline src/utils/task_stats_alloc/src/lib.rs:77]: I think this is the source of contention. Why we need to check it every time we sub it? We can have a background task to hold the counter when task finished.

Comment (BowenXiao1999) [reply src/utils/task_stats_alloc/src/lib.rs:77]: I will remove this and bench it again. 

Actually I also want to test hytra, but because hytra only supports i64 but not u64, not sure whether there will be some problems (guess no).

File: src/utils/task_stats_alloc/src/lib.rs (hunk 8 above)
Comment (fuyufjh) [reply src/utils/task_stats_alloc/src/lib.rs:147]: My fault. Misunderstood the doc. 😇

File: src/utils/task_stats_alloc/src/lib.rs (hunk 9 above)
Comment (liurenjie1024) [reply src/utils/task_stats_alloc/src/lib.rs:77]: If we still use current approach, I don't think hytra will help much, but you can have a try.

**Comment by BowenXiao1999** [comment]
Perf Update, pick some numbers to show roughly:
Sysbench (primary key point lookup): 
qps: 
151763 (before), 161062 (after), 169699 (main)

p99 latency:
2.11 (before), 2 (after), 2.03 (main)

Nexmark q7:
source throughput (max):
955210 (before), 1370000 - 1410000 (after), 1570000 (main)

node memmory:
7.45GB (before), 9.31GB (after), 9.31GB (main)

Basically from these rough numbers we can see that, after adopt the atomic order change, the performance loss seems acceptable, and memory usage do not go up. Looks good. And maybe in future we can add more optimizations if possible? (e.g. replace AtomicSize)

I will use [loom](https://github.com/tokio-rs/loom) to see whether it's ok to add some test.

**Comment by fuyufjh** [comment]
> 1. Always use Ordering::Relax
> 2. remove the fence

@BowenXiao1999 I remembered that several months ago, @BugenZhao and @TennyZhuang had done some test on this branch and fixed several "leaks", which means the Atomic counter doesn't reduce to zero when all objects were dropped, so the counter itself cannot be recycled forever. Can you confirm with them whether these `Ordering` and fencing stuff are related?

File: src/utils/stats_alloc/src/task_local.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (BowenXiao1999) [reply src/utils/stats_alloc/src/task_local.rs]: Good catch @wangrunji0408 ! Just want to make sure my understanding.

Given that fence(Ordering::Acquire) here is unnecessary, Do you also think that  fetch_sub with Ordering::Release is also unnecessary -> we can replace all orders to be Ordering::Relaxed?

[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/utils/stats_alloc/src/task_local.rs]: I'm afraid we can't. May refer to the discussions here.

https://www.boost.org/doc/libs/1_55_0/doc/html/atomic/usage_examples.html#boost_atomic.usage_examples.example_reference_counters

**Comment by BugenZhao** [comment]
The memory fence path is only covered if we're to drop a counter, which only happens if we drop a task. In streaming, this is an infrequent behavior. So I wonder whether the results are convincing enough. 🥺

File: src/utils/stats_alloc/src/task_local.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (BowenXiao1999) [reply src/utils/stats_alloc/src/task_local.rs]: Thanks for the Strong statement! Not sure which part you want to argue, 1) we should keep original acquire-release fence 2) we should not use fence but use the Release Order (?)

But seems the boost ref counter example has something different from our case: IIUC, the refcount has object and counter, while here we only get counter (itself is the object we want to delete). So as @wangrunji0408 mention, seems like this is an exeception?

**Comment by BowenXiao1999** [comment]
> The memory fence path is only covered if we're to drop a counter, which only happens if we drop a task. In streaming, this is an infrequent behavior. So I wonder whether the results are convincing enough. 🥺

Good catch. To find the cause, I plan to compare the metrics more deeply. Everyone interesting can also have a try on this bench, more tests mean less uncertainty.

> BTW, we don't have this monitoring enabled in batch tasks in this PR, so the only overhead for batch query is calculating the offset and put a NULL pointer to show that this memory is not tracked. Why does changing the counter logic affect the batch query performance?

Yes, batch is not enabled monitoring (I mean reported as metrics). I just guess, is it because batch task usually will decrease to 0 and the usize stuff. monitoring is just a background task so I thought the main overhead comes from Allocator itself? 🤔

File: src/utils/stats_alloc/src/task_local.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/utils/stats_alloc/src/task_local.rs]: Yes, I believe the `Release` in above `fetch_sub` can also be replaced by `Relaxed` safely.

**Comment by wangrunji0408** [comment]
> The memory fence path is only covered if we're to drop a counter, which only happens if we drop a task. In streaming, this is an infrequent behavior. So I wonder whether the results are convincing enough. 🥺

I guess what matters is the `fetch_sub(val, Ordering::Release)`. It happens every time `dealloc` is called, and will flush the write buffer inside the CPU core, which hurts the performance.

**Comment by TennyZhuang** [comment]
I will try another solution and test the performance.

**Comment by BowenXiao1999** [comment]
Shall we merge this pr and add #6822 later?

**Comment by liurenjie1024** [comment]
> Shall we merge this pr and add #6822 later?

LGTM

File: src/utils/task_stats_alloc/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+#[inline(always)]
+fn wrap_layout(layout: Layout) -> (Layout, usize) {
+    debug_assert_ne!(layout.size(), 0, "the size of layout must be non-zero");
+
+    let (wrapped_layout, offset) = Layout::new::<TaskLocalBytesAllocated>()
+        .extend(layout)
+        .expect("wrapping layout overflow");
+    let wrapped_layout = wrapped_layout.pad_to_align();
```
Comment (liurenjie1024) [inline src/utils/task_stats_alloc/src/lib.rs:131]: Add a check here to ensure new alignment matches old alignment? cc @wangrunji0408 @BowenXiao1999

Comment (liurenjie1024) [reply src/utils/task_stats_alloc/src/lib.rs:131]: Seems that `extend` method takes it into account.

**Review by liurenjie1024 — APPROVED** [review]
LGTM

File: src/utils/task_stats_alloc/src/lib.rs (hunk 10 above)
Comment (wangrunji0408) [reply src/utils/task_stats_alloc/src/lib.rs:131]: What exact variables of 'old' and 'new' do you mean?

If you mean this line:
```rust
let wrapped_layout = wrapped_layout.pad_to_align();
//  ^ new            ^ old
```
The alignment won't be changed by definition and by its implementation.

If you mean the function:
```rust
fn wrap_layout(layout: Layout) -> (Layout, usize)
//             ^ old               ^ new
```
I guess it is possible to be changed?

Comment (liurenjie1024) [reply src/utils/task_stats_alloc/src/lib.rs:128]: The later one. But according to comments in `extend` methods, I think old layout's alignment is still respected.

File: src/utils/task_stats_alloc/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    #[inline(always)]
+    unsafe fn add_unchecked(&self, val: usize) {
+        self.0.unwrap_unchecked().fetch_add(val, Ordering::Relaxed);
+    }
+
+    /// Subtracts from the counter value, and `drop` the counter while the count reaches zero.
+    #[inline(always)]
+    fn sub(&self, val: usize) {
+        if let Some(bytes) = self.0 {
+            // Use Relax as we only need to make it atomic.
```
Comment (wangrunji0408) [inline src/utils/task_stats_alloc/src/lib.rs:67]: A little suggestion for the comment:
```suggestion
            // Use `Relaxed` order as we don't need to sync read/write with other memory addresses.
            // Accesses to the counter itself are serialized by atomic operations.
```

File: src/utils/task_stats_alloc/src/lib.rs (hunk 10 above)
Comment (wangrunji0408) [reply src/utils/task_stats_alloc/src/lib.rs:128]: But if the old layout comes from a `u8` (size = 1, align = 1), the new layout should be `(usize, u8)` (size = 16, align = 8)?
